File: src/main/java/io/koju/autopos/shared/domain/AbstractEntity.java
package io.koju.autopos.shared.domain;

import org.springframework.util.ClassUtils;

import javax.persistence.MappedSuperclass;
import javax.persistence.Transient;

@MappedSuperclass
public abstract class AbstractEntity implements Persistable<Long> {

    private static final long serialVersionUID = -5554308939380869754L;

    protected abstract void setId(final Long id);

    @Transient
    public boolean isNew() {
        return null == getId();
    }


    @Override
    public String toString() {
        return String.format("Entity of type %s with id: %s", this.getClass().getName(), getId());
    }


    @Override
    public boolean equals(Object obj) {

        if (null == obj) {
            return false;
        }

        if (this == obj) {
            return true;
        }

        if (!getClass().equals(ClassUtils.getUserClass(obj))) {
            return false;
        }

        AbstractEntity that = (AbstractEntity) obj;

        return null == this.getId() ? false : this.getId().equals(that.getId());
    }


    @Override
    public int hashCode() {

        int hashCode = 17;

        hashCode += null == getId() ? 0 : getId().hashCode() * 31;

        return hashCode;
    }
}

File: src/main/kotlin/com/kapilkoju/autopos/kernel/domain/AbstractEntity.kt
package com.kapilkoju.autopos.kernel.domain

import com.fasterxml.jackson.annotation.JsonIgnore
import org.springframework.util.ClassUtils
import java.io.Serializable
import javax.persistence.MappedSuperclass
import javax.persistence.Transient

@MappedSuperclass
abstract class AbstractEntity<out ID : Serializable> : IdentifiablePersistable<ID> {

    @Transient
    @JsonIgnore
    override fun isNew(): Boolean {
        return null == getIdentity()
    }


    override fun toString(): String {
        return String.format("Entity of type %s with identity: %s", this.javaClass.name, getIdentity())
    }


    override fun equals(other: Any?): Boolean {

        if (null == other) {
            return false
        }

        if (this === other) {
            return true
        }

        if (javaClass != ClassUtils.getUserClass(other)) {
            return false
        }

        val that = other as AbstractEntity<*>?

        return if (null == this.getIdentity()) false else this.getIdentity() == that!!.getIdentity()
    }


    override fun hashCode(): Int {

        var hashCode = 17

        hashCode += if (null == getIdentity()) 0 else getIdentity()!!.hashCode() * 31

        return hashCode
    }

    companion object {

        private val serialVersionUID = -5554308939380869754L
    }
}

